apps/designer/app/canvas/features/text-editor/interop.ts

```diff
@@ -0,0 +1,100 @@
+import {
+  type ElementNode,
+  $getRoot,
+  $createTextNode,
+  $createParagraphNode,
+  TextNode,
+  LineBreakNode,
+  ParagraphNode,
+} from "lexical";
+import { $createLinkNode, LinkNode } from "@lexical/link";
+import type { ChildrenUpdates, Instance } from "@webstudio-is/react-sdk";
+import { createInstanceId } from "~/shared/tree-utils";
+
+// Map<nodeKey, Instance>
+export type Refs = Map<string, Instance>;
+
+const $writeUpdates = (
+  node: ElementNode,
+  updates: ChildrenUpdates,
+  refs: Refs
+) => {
+  const children = node.getChildren();
+  for (const child of children) {
+    if (child instanceof ParagraphNode) {
+      if (updates.length !== 0) {
+        updates.push("\n");
+      }
+      $writeUpdates(child, updates, refs);
+    }
```

Comment:
this would make it clear that we aren't going to do anything later, so we can stop reading
```suggestion
      continue;
    }
```